Major Refactor

timgerstel opened the pull request:
1. Make sure yaml and schema are ALWAYS set
2. Change the description of smpe on the "Installation Type" stage as per @struga0258 's suggestion
5. Wrap all global variables in components in useState() to reduce memory cost *Edit: this is still here but it was actually useless lol
6. Fix scrolling on most stages
7. Fix Initialization stage status not updating when all substeps are complete
8. Add popup messages for upload yaml button on Launch Config/Networking stages
9. Make editor use entire Zowe schema
10. Use more Constants.tsx
11. Add ability to retrieve example-zowe.yaml and JSON schemas when using SMPE mode
12. Remove duplicate and unused vars
13. Fix cursor jumping to end of port input on networking stage
14. "Skip" on Unpax stage now fetches example-zowe.yaml and schemas
15. Integrate @skurnevich's method of downloading example-zowe.yaml and schemas instead of parsing cat commands🐈
16. Fix upload not working
17. Fix improper merging of example-zowe.yaml and current config thanks to @skurnevich deep merge method
18. Remove unused imports
19. Reduce duplication
20. Remove unused or unnecessary function args
21. Fix schema errors
22. Fix difference between smpe install dir and download/upload. Download/upload will no longer unpax to /runtime
23. Fix makeDir infinite loop on planning stage with invalid dirs
24. Fix yaml encoding so that zwe start can be run
25. Make editor wider
26. Fix certificates stage resetting zowe.setup.certificate values when changing zowe.verifyCertificates 
27. Increase editor width and height
28. Probably more i forgot

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Change in a documentation
- [ ] Refactor the code 
- [ ] Chore, repository cleanup, updates the dependencies.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## PR Checklist
Please delete options that are not relevant.
- [ ] If the changes in this PR are meant for the next release / mainline, this PR targets a "staging" branch.
- [ ] My code follows the style guidelines of this project (see: [Contributing guideline](https://github.com/zowe/zen/blob/v2.x/staging/CONTRIBUTING.md))
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] New and existing unit tests pass locally with my changes
- [ ] video or image is included if visual changes are made
<!-- - [ ] Relevant update to CHANGELOG.md -->
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works, or describe a test method below

## Testing
<!-- Describe how this code should be tested. I've you've added an automated or unit test, then describe how to run it. Otherwise, describe how you have tested it and how others should test it. -->

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, or if there are follow-up tasks and TODOs etc... -->

timgerstel committed:
Updating editor no longer crashes Zen. Updating components now updates UI properly

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
More robust conditional

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Update stage status when uploading yaml

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Ideally this would be its own separate component but overhead with the editor was making things too tedious given the deadline

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Disallow non numeric input

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Merge branch 'v2.x/staging' into bugfix/externalDomainsAndYaml

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Add error notification for failed yaml upload. Set state false if fail

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove console log

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Bulk commit because i need to check staging for something

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
No longer set schema and config on EVERY single stage. Schema, config, and installation args are now set on the Home component. Wrap naked component global vars in useState calls. Fix auto scrolling when a stage has been complete

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Better AJV implementation

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
use useState

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
useState

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Update verbage

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Update initialization status properly when all substages are complete

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Reset stage status when uploading yaml

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix crash on first run of Zen, fix installation args not setting peroperly, remove unneeded ipcRenderer hooks

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Original fallback schema is needed because of AJVs inability to resolve refs properly

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix schema errors

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix init apf auth weird scrolling

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix launch config status resetting

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix edge case of zowe.job not being set??? shouldnt be possible

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix installation type not being saved

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
remove console logs

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Use correct installation args

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Add button to retrieve example-zowe.yaml and schemas when using SMPE mode

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Always use installation args

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Rewrite all of pattern properties so port input doesnt jump to end lol

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

skurnevich commented:
Does not work for me:

Local pax file uploading fails
<img width="1348" alt="Screenshot 2024-06-13 at 14 34 53" src="https://github.com/zowe/zen/assets/61471197/f735a6cb-78a5-47a2-b056-da221d122761">


While downloading pax fails on schema parsing
```
14:43:42.429 › error setting schema from pax: SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
```

For the last one i have created the [PR 179](https://github.com/zowe/zen/pull/179) to replace cat command output parsing with files downloading. 
You can merge it first, or incorporate changes into this PR. The factual change is smaller than it looks, you just need to replace readExampleYamlAndSchema function and remove parseExampleYamlFromPax blocks as downloaded yaml and json files are ready to use

skurnevich commented on src/actions/InstallationHandler.ts:
```diff
@@ -140,64 +272,17 @@ class Installation {
               // console.log("current config:", JSON.stringify(currentConfig));
               // console.log("yamlObj: ", JSON.stringify(yamlObj));
-              yamlObj = {...currentConfig, ...yamlObj}
+              yamlObj = yamlObj = Object.assign({}, yamlObj, currentConfig);
```
Looks like a typo

skurnevich commented on src/actions/InstallationHandler.ts:
```diff
@@ -140,64 +272,17 @@ class Installation {
-            if (yamlObj.zOSMF.applId === undefined && installationArgs.zosmfApplId) {
-              yamlObj.zOSMF.applId = installationArgs.zosmfApplId;
-            }
+            this.mergeYamlAndInstallationArgs(yamlObj, installationArgs);
```
Is there a reason why we do this merge with installationArgs values one by one, if it should have been done a line above with yamlObj = Object.assign({}, yamlObj, currentConfig); 
Is there a case when currentConfig will not exist?

timgerstel replied:
At one point, during the very first run of Zen current config did not exist. I haven't confirmed this behavior in a while but in my head i dont think it should ever not exist currently.

skurnevich commented on src/actions/InstallationHandler.ts:
```diff
@@ -50,12 +51,143 @@ class Installation {
+    }
+  }
+
+  public async smpeGetExampleYamlAndSchemas (
```
What is different about smpe installation schemas to have this code separately?

timgerstel replied:
Very good point, i changed it to reuse this function in the downloadUnpax function to reduce duplication

timgerstel committed:
Fix next button:

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

skurnevich commented on src/renderer/components/stages/Certificates.tsx:
```diff
@@ -26,28 +26,24 @@ import { createTheme } from '@mui/material/styles';
-  const STAGE_ID = getStageDetails(stageLabel).id;
-  const SUB_STAGES = !!getStageDetails(stageLabel).subStages;
-  const SUB_STAGE_ID = SUB_STAGES ? getSubStageDetails(STAGE_ID, subStageLabel).id : 0;
+  const [STAGE_ID] = useState(getStageDetails(INIT_STAGE_LABEL).id);
```
Question on this, we should use useState for variables which we want our component to be re-rendered on change.  Why use it for constants that aren't change in scope of this component?

timgerstel replied:
> Why use it for constants that aren't change in scope of this component?

If it's not wrapped in useState, react will reinitialize the variable even when the component doesnt rerender I believe. If you put a naked console.log in the global scope of a React component, it will keep getting called over and over. This leak could get costly on memory so I play it safe and wrap everything in useState.

I could also be mistaken about the re-initialization part

timgerstel replied:
I think I was mistaken about these variables needing to be in useState

DivergentEuropeans replied:
I haven't tested it, but I think @timgerstel might be right here. Not that you're wrong @skurnevich , I think both things are true: using const with useState introduces a small overhead. So it would be like:
For constants and simple values that do not involve expensive computation, use them directly without useState.
For values that involve expensive computation or need to persist across renders, use useState.

So something like this might be max/minning:
```
const Certificates = () => {
  const theme = useState(createTheme())[0]; // Expensive
  const stageLabel = 'Initialization';
  const STAGE_ID = useState(getStageDetails(INIT_STAGE_LABEL).id)[0]; // Potentially expensive?
};
```

skurnevich replied:
No worries, @DivergentEuropeans, that was more like a question from my side. 
So you say that in this case:
`const [STAGE_ID] = useState(getStageDetails(INIT_STAGE_LABEL).id)`
`getStageDetails` function will be triggered less times than in this case: 
`const STAGE_ID = getStageDetails(INIT_STAGE_LABEL).id`
Right? Or it will bring some other performance changes? 

I am asking because i've never seen such usage before. 
For caching really expensive operations i get used to use "useMemo"

skurnevich commented on src/renderer/components/stages/Networking.tsx:
```diff
@@ -18,636 +18,545 @@ import Ajv from "ajv";
-            "minimum": 0,
-            "maximum": 65535,
-            "description": "Port number of how you access Zowe APIML Gateway from your local computer."
+const schema: any = {
```
What about this schema, it is zowe-yaml-schema.json right? if we need if for fallback then probably we can move it out to constants, or have fallback schemas just shipped as files with zen?

timgerstel replied:
This schema is actually different than zowe-yaml-schema.json, it hardcodes references from the server-common.json schema because AJV was having trouble resolving references. It also has a few fixes for regexes (zoweDataset, zoweDatasetMember) that AJV did not like

DivergentEuropeans replied:
So if BASE_SCHEMA is AJV "fixed", can't we remove the schemas in Networking + Launch Config and use BASE_SCHEMA? Or lmk I might be missing something

timgerstel committed:
Dont extract files to installationDir/runtime, just extact to installationDir

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
fix wrong dir for unpax

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
fix unpax relative path again lol

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
add missing bin part of dir

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Dont change schema id

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove unused var

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
set additional properties true because schema seems to need updating

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix externaldomains non existent

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix scrolling on apf auth

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix vsam page not updating yaml

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Retrieve example-yaml and schemas when skipping unpax if runtime files already exist

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix typo

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Implement sergei's file downloading method for fetching example-zowe.yaml and schemas

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Each stage will only compile its own part of the schema with ajv. Fix order of operations for object.assign call when merging example-zowe.yaml

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Reduce duplication of example-yaml + schema fetch

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel commented:
@skurnevich 

Thank you so much for your review.

>For the last one i have created the https://github.com/zowe/zen/pull/179 to replace cat command output parsing with files downloading.
You can merge it first, or incorporate changes into this PR.

I implemented this code and I love it so much more than doing string manipulations which were unreliable due to the job output not always having the same format for some reason. I implemented it into this PR.

timgerstel committed:
userUploadedPaxPath not persisting for some reason???

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove unused imports, wrap more global component vars in useState, fix upload path

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fetch example-zowe.yaml and schemas when skipping upload as well

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

skurnevich commented on src/actions/InstallationHandler.ts:
```diff
@@ -44,18 +46,127 @@ class Installation {
+          try {
+            yamlObj = parse(yamlFromPax);
+            if (currentConfig) {
+              yamlObj = Object.assign({}, yamlObj, currentConfig);
```
This bug comes from here
`18:45:09.025 › error parsing example-zowe.yaml: TypeError: Cannot set properties of undefined (setting 'name')`
Object.assign replaces complete zowe section so some default properties disappear, we should use some deep merge algorithm, we can use some external lib, but i would recommend to avoid adding more dependencies and would just use some recursive func like this:
```
function deepMerge(base, extension) {
  if (isObject(base) && isObject(extension)) {
    for (const key in extension) {
      if (isObject(extension[key])) {
        if (!base[key]) base[key] = {};
        deepMerge(base[key], extension[key]);
      } else {
        Object.assign(base, {[key]: extension[key]});
      }
    }
  }
  return base;
}
```

Also it worth to review other places in the code where we merge several configs...

timgerstel replied:
Added this, another great suggestion. Really appreciate your review on this massive PR!

timgerstel committed:
Fix name undefined during unpax

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove unused args from init functions. Remove getConfig call from ipcRenderer hook call and simply move it to the function within InstallationHandler

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Deep merge of current config and example-zowe.yaml, big thank you to Sergei

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix typo

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remveo consoe log

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Add spacing to pattern match titles

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Merge remote-tracking branch 'origin/v2.x/staging' into fullRewrite

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix typos

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Add error to log message

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix setting of installationArgs

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Update networking status to false on change

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix scrolling on apf auth stage

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Scrolling fixes

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove unused imports

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove unused imports

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix ui error when typing invalid dir on planning

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Merge remote-tracking branch 'origin/v2.x/staging' into fullRewrite

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Return false on error

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Set state false when validating locations:

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix makeDir infinite loop with invalid dir

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Add error message for failure

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Use correct installation args

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove installationArgs from editor

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix verify certs not updating in yaml

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Add error popup to certificates stage

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Format output

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix typo, this might not even make the outputl ook good anyway

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fetch updated yaml after init certificate

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Set updated yaml from --update-config flag if it exists

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Deep merge object

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix runtime error

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Update import

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Move updatedYaml logic to correct location

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Update init status on app start

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Unset because New Installtion is supposed to reset the state anyway so resume probably handles this??

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Proper merging

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove yaml downloading

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix yaml download logic

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Move parsing into try catch

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Do not store password

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Reset state on new install

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove comment but i still believe this setter to be redundant

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Merge remote-tracking branch 'origin/v2.x/staging' into fullRewrite

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Better error handling for init certificate failure

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Better error handling for init mvs

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
certificate -> mvs

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Add error handling for init apfauth, remove redundant console.warn

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Better error handling for security

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Reset stc stage on new installation

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Better error handling for init stc

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Better error handling for init vsam

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Merge branch 'v2.x/staging' into fullRewrite

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Update tag script for yaml

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
use filePath

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Does it need a tag first??

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix status resetting when going back from a successful install

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix apf auth status not resetting on re init

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Move status reset logic

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Why dont any of these onClicks work

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix tagging script

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Better conditional for showing error

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Properly update certificate section after init certificate

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Update local yaml as well

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Yaml still wasnt updating properly in zen:

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Flip order of base/extension?

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix yaml setting issue

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Deep merge caused issue w array to yaml conversion

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Set local yaml to wrong section but doesnt even need to be set

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

skurnevich commented:
Hi @timgerstel is this PR ready for review/merge or would you like to add more changes?

DivergentEuropeans committed:
Revert create recursive dirs + add edgecase

Signed-off-by: Leanid Astrakou <lastrakou@rocketsoftware.com>

DivergentEuropeans commented:
Created PR: https://github.com/zowe/zen/pull/194

DivergentEuropeans committed:
Added a stop recursion flag for bad permissions edgecase

Signed-off-by: Leanid Astrakou <lastrakou@rocketsoftware.com>

timgerstel committed:
Merge pull request #194 from zowe/fix-recursive-dirs

Revert create recursive dirs + add edgecase

timgerstel committed:
Only create runtime dir

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix typo

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix yaml being out of sync on cert stage

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Reset certificates status on input change

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Remove status checks for workspace, log, extension dir as zen will not longer attempt to create these

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
New installation now resets state properly

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Dont load last state if isResume

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Increase editor width and height

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Merge branch 'v2.x/staging' into fullRewrite

timgerstel committed:
Reset values on new install

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

DivergentEuropeans commented on src/main/index.ts:
```diff
@@ -195,23 +194,28 @@ const createWindow = (): void => {
-  ipcMain.handle('download-unpax', async (event, connectionArgs, installationArgs, version, zoweConfig) => {
-    const res = await installActions.downloadUnpax(connectionArgs, installationArgs, version, zoweConfig);
+  ipcMain.handle('get-yaml-schema', async (event, connectionArgs, installationArgs) => {
+    const res = await installActions.smpeGetExampleYamlAndSchemas(connectionArgs, installationArgs);
```
shouldn't this be called "getExampleYamlAndSchemas"

I only see this used in 2 places, 1 in Stepper which is conditional to not SMPE and 2 in 
Unpax step which is also not SMPE

timgerstel replied:
sure

DivergentEuropeans commented on src/renderer/components/Home.tsx:
```diff
@@ -58,12 +57,24 @@ const cards: Array<ICard> = [
-        <Card id={`card-${id}`} square={true} >
+    <Link key={`link-${id}`} to={link} >
+      <Box sx={{ width: '40vw', height: '40vh'}} onClick={(e) => {
+        const flattenedData = flatten(lastActiveState);
```
`lastActiveState `is already a flat object with no nested properties, at a quick glance, don't see it set anywhere. Is this essential?

timgerstel replied:
yeah youre right but that's pretty minor i dont that one time call when the user clicks "new installation" would cause a noticeable performance impact

DivergentEuropeans commented on src/renderer/components/Home.tsx:
```diff
@@ -109,6 +116,42 @@ const Home = () => {
 
 
+
+    //Home is the first screen the user will always see 100% of the time. Therefore, we will call the loading of the configs, schemas, and installation args here and set them to the redux memory states
```
Makes sense 👍

DivergentEuropeans replied:
Home.tsx change is v satisfying

DivergentEuropeans commented on src/renderer/components/common/EditorDialog.tsx:
```diff
@@ -82,23 +79,18 @@ const EditorDialog = ({contentType, isEditorVisible, toggleEditorVisibility, onC
     // To validate the javascript object against the schema
-    const isValid = validate(jsonData);
-    setIsSchemaValid(isValid);
+    setIsSchemaValid(!schemaValidate.errors);
```
N i c e

DivergentEuropeans replied:
that all that adjv stuff worked

DivergentEuropeans commented on src/renderer/components/stages/Certificates.tsx:
```diff
@@ -63,50 +57,10 @@ const Certificates = () => {
-  if(verifyCertsSchema) {
-    validateVerifyCertSchema = ajv.compile(verifyCertsSchema);
-  }
+  const [validate] = useState(() => ajv.getSchema("https://zowe.org/schemas/v2/server-base") || ajv.compile(setupSchema))
```
V nitpick but all of these `https://zowe.org/schemas/v2/server-base` should be Constants.BASE_SCHEMA.$id

timgerstel replied:
they could all be replaced yes

DivergentEuropeans commented on src/renderer/components/stages/Certificates.tsx:
```diff
@@ -205,13 +159,28 @@ const Certificates = () => {
       clearInterval(timer);
-    }).catch(() => {
+      updateProgress(res.status);
+      if(!res.status){
```
This may be unfinished. Not that it's relevant to this PR, it's already a large and good one, but in other steps to correctly determine if we need toggleEditor, we need checks for script failure status, res.status is unreliable. Something like,

```if (res?.details && res.details.jobOutput.indexOf(JCL_UNIX_SCRIPT_OK) == -1) {```

Not sure if you had a chance to test all those res.status cases. I would have res.status = 0, but the command failed. TODO for a later PR could be fine

DivergentEuropeans commented on src/renderer/components/stages/InitApfAuth.tsx:
```diff
@@ -236,19 +196,16 @@ const InitApfAuth = () => {
-        } else {
-          console.warn('zwe init apfauth failed', err?.toString()); // toString() throws run-time error on undefined or null
-        }
+        window.electron.ipcRenderer.setStandardOutput(`zwe init apfauth failed:  ${typeof err === "string" ? err : err.toString()}`).then((res: any) => {
```
If this was tested, then you want to add:
`alertEmitter.emit('showAlert', 'Please view Job Output for more details', 'error');`

timgerstel replied:
That is redundant because the editor automatically pops up we don't need a popup to tell the user to look at the output that they are already looking at

DivergentEuropeans commented on src/renderer/components/stages/Networking.tsx:
```diff
@@ -787,7 +684,75 @@ const Networking = () => {
+              }
+               return null;
+            })}
+          </div>
```
This whole chunk is difficult to rea dfor me, but it looks mostly like a "code move" so it seems alright

timgerstel committed:
Fix error for undefined object

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

DivergentEuropeans committed:
Reset counter properly + don't forget the current char

Signed-off-by: Leanid Astrakou <lastrakou@rocketsoftware.com>

DivergentEuropeans committed:
Remove comments

Signed-off-by: Leanid Astrakou <lastrakou@rocketsoftware.com>

timgerstel committed:
Merge pull request #198 from zowe/bugfix/script-parsing

Reset counter properly + don't forget the current char

DivergentEuropeans committed:
Another small edgecase

Signed-off-by: Leanid Astrakou <lastrakou@rocketsoftware.com>

timgerstel committed:
Merge pull request #199 from zowe/bugfix/script-parsing

Another small edgecase

DivergentEuropeans commented on src/renderer/components/stages/Security.tsx:
```diff
@@ -234,19 +189,17 @@ const Security = () => {
-      } else {
-        console.warn('zwe init security failed', err?.toString()); // toString() throws run-time error on undefined or null
-      }
+      window.electron.ipcRenderer.setStandardOutput(`zwe init security failed:  ${typeof err === "string" ? err : err.toString()}`).then((res: any) => {
```
If this was tested, then you want to add:
alertEmitter.emit('showAlert', 'Please view Job Output for more details', 'error');

timgerstel replied:
>So if BASE_SCHEMA is AJV "fixed", can't we remove the schemas in Networking + Launch Config and use BASE_SCHEMA? Or lmk I might be missing something

No because the networking stage had some caveats that required its own schema which is modified from the one packaged with Zowe, maybe it can now be amended ill have to look into it

DivergentEuropeans commented on src/renderer/components/stages/installation/Installation.tsx:
```diff
@@ -271,12 +255,12 @@ const Installation = () => {
           res.details = { 3: res.details };
         }
-        if (res?.details && res.details[3] && res.details[3].indexOf(JCL_UNIX_SCRIPT_OK) == -1) { // This check means we got an error during zwe install
+        if (!res.status && res?.details && res.details[3] && res.details[3].indexOf(JCL_UNIX_SCRIPT_OK) == -1) { // This check means we got an error during zwe install
```
I don't think this is a good addition. I remember cases where res.status = 0, yet the command failed. That was the whole point of using 'JCL_UNIX_SCRIPT_OK' as a better source of truth over status

DivergentEuropeans commented on src/renderer/components/stages/Unpax.tsx:
```diff
@@ -90,6 +89,29 @@ const Unpax = () => {
+    setDownloadUnpaxProgress(downloadUnpaxStatus);
+    dispatch(setNextStepEnabled(false));
+    window.electron.ipcRenderer.fetchExampleYamlBtnOnClick(connectionArgs, installationArgs).then((res: IResponse) => {
+      if(!res.status){ //errors during runInstallation()
```
Same point here about unreliability of res.status

timgerstel committed:
Fix undefined port

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Add worpWrap

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Fix job output, mostly

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Dont store password

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

timgerstel committed:
Only wordwrap non yaml output

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>
